src/pkg/provisioner/state.go

```diff
@@ -52,7 +60,58 @@ func (s *state) write() error {
 	return nil
 }
 
-func initState(dir string, c Config) (*state, error) {
+func downloadGCSObject(ctx context.Context, gcsClient *storage.Client, bucket, object, localPath string) error {
+	address := fmt.Sprintf("gs://%s/%s", bucket, object)
+	gcsObj, err := gcsClient.Bucket(bucket).Object(object).NewReader(ctx)
+	if err != nil {
+		return fmt.Errorf("error reading %q: %v", address, err)
+	}
+	defer utils.CheckClose(gcsObj, fmt.Sprintf("error closing GCS reader %q", address), &err)
+	localFile, err := os.Create(localPath)
+	if err != nil {
+		return err
+	}
+	defer utils.CheckClose(localFile, "", &err)
+	if _, err := io.Copy(localFile, gcsObj); err != nil {
+		return fmt.Errorf("error copying %q to %q: %v", address, localFile.Name(), err)
+	}
+	return nil
+}
+
+func (s *state) unpackBuildContexts(ctx context.Context, deps Deps) (err error) {
+	for name, address := range s.data.Config.BuildContexts {
+		log.Printf("Unpacking build context %q from %q", name, address)
+		if address[:len("gs://")] != "gs://" {
+			return fmt.Errorf("cannot use address %q, only gs:// addresses are supported", address)
+		}
+		splitAddr := strings.SplitN(address[len("gs://"):], "/", 2)
+		if len(splitAddr) != 2 || splitAddr[0] == "" || splitAddr[1] == "" {
+			return fmt.Errorf("address %q is malformed", address)
+		}
+		bucket, object := splitAddr[0], splitAddr[1]
+		tarPath := filepath.Join(s.dir, name+".tar")
+		if err := downloadGCSObject(ctx, deps.GCSClient, bucket, object, tarPath); err != nil {
+			return fmt.Errorf("error downloading %q to %q: %v", address, tarPath, err)
+		}
+		tarDir := filepath.Join(s.dir, name)
+		if err := os.Mkdir(tarDir, 0770); err != nil {
+			return err
+		}
+		args := []string{"xf", tarPath, "-C", tarDir}
```

Comment:
should some of these args come from `Deps.Tarcmd` somehow? or is the `Tarcmd` basically allowing supply of something like an alias to tar?

Reply:
TarCmd is for providing a tar alias, yeah. It's helpful for testing, if you don't want to depend on the tar implementation in the environment. For now, I think the interface is cleaner without allowing for passing additional args, but we can change it later if it makes sense later.